### PR TITLE
Fixed the PR from (dirigeants/core#263)

### PR DESCRIPTION
**Pull request by Relms12345**
Same code but I fixed it to prevent the build fails

### CODE REVIEW

File: src/lib/caching/structures/guilds/Guild.ts [hunk 1]
```diff
@@ -488,6 +488,14 @@ export class Guild extends Structure {
 		return this.client.api.get(Routes.guildVanityURL(this.id)) as Promise<GuildVanityURL>;
 	}
 
+	/**
+	 * Returns the icon url for the guild if one is available.
+   * @param options The image size and format options.
+   */
```
Comment (bdistin) [inline src/lib/caching/structures/guilds/Guild.ts:494]:
```suggestion
	/**
	 * Returns the icon url for the guild if one is available.
	 * @param options The image size and format options.
	 */
```

File: src/lib/caching/structures/User.ts [hunk 2]
```diff
@@ -138,6 +138,24 @@ export class User<T = Client> extends Structure<T> {
...
+	}
+
+	/**
+	 * Returns the users avatar url or the default discord avatar url if they don't have a avatar.
+	 * @param options The image size, format and other options.
+	 */
+	public displayAvatarURL(options?: ImageURLOptions): string | null {
+		if (!(this.client instanceof Client)) return null;
+		return this.avatarURL(options) || this.client.api.cdn.defaultAvatar(Number(this.discriminator));
+	}
```
Comment (kyranet) [inline src/lib/caching/structures/User.ts:157]:
```suggestion
	public displayAvatarURL(options?: ImageURLOptions): T extends Client ? string : null {
		if (!(this.client instanceof Client)) return null as T extends Client ? string : null;
		return (this.avatar ? 
			this.avatarURL(options) :
			this.client.api.cdn.defaultAvatar(Number(this.discriminator))) as T extends Client ? string : null;
	}
```

File: src/lib/caching/structures/guilds/GuildMember.ts [hunk 3]
```diff
@@ -212,6 +212,22 @@ export class GuildMember extends Structure {
...
+	public avatarURL(options?: ImageURLOptions): string | null {
+		return this.user?.avatar ? this.client.api.cdn.userAvatar(this.id, this.user.avatar, options) : null;
+	}
+
+	/**
+	 * Returns the users avatar url or the default discord avatar url if they don't have a avatar.
+	 * @param options The image size, format and other options.
+	 */
+	public displayAvatarURL(options?: ImageURLOptions): string {
+		return this.avatarURL(options) || this.client.api.cdn.defaultAvatar(Number(this.user?.discriminator));
```
Comment (kyranet) [inline src/lib/caching/structures/guilds/GuildMember.ts:228]:
```suggestion
		return this.avatarURL(options) ?? this.client.api.cdn.defaultAvatar(Number(this.user?.discriminator));
```

File: src/lib/caching/structures/oauth/Application.ts [hunk 4]
```diff
@@ -129,6 +129,15 @@ export class Application {
 		return this.guildID !== null ? this.client.guilds.get(this.guildID) ?? null : null;
 	}
 
+	/**
+	 * Returns the application icon url if available.
+	 * @param options The image size, format, and other image url options.
+	 */
+	public iconURL(options?: ImageURLOptions): string | null {
+		if (!this.icon) return null;
+		return this.client.api.cdn.appIcon(this.id, this.icon, options);
```
Comment (kyranet) [inline src/lib/caching/structures/oauth/Application.ts:138]:
```suggestion
		return this.icon ? this.client.api.cdn.appIcon(this.id, this.icon, options) : null;
```

File: src/lib/caching/structures/oauth/Team.ts [hunk 5]
```diff
@@ -48,4 +49,13 @@ export class Team {
 		return this.members.get(this.ownerID) as TeamMember;
 	}
 
+	/**
+	 * Returns the teams icon url if available.
+	 * @param options The image size, format, and other image url options.
+	 */
+	public iconURL(options?: ImageURLOptions): string | null {
+		if (!this.icon) return null;
+		return this.client.api.cdn.teamIcon(this.id, this.icon, options);
```
Comment (kyranet) [inline src/lib/caching/structures/oauth/Team.ts:58]:
```suggestion
		return this.icon ? this.client.api.cdn.teamIcon(this.id, this.icon, options) : null;
```

**Comment by bdistin** [comment]
Please fix the trailing space

**Comment by Relms12345** [comment]
@bdistin fixed

**Comment by kyranet** [comment]
`userAvatar`:
https://github.com/dirigeants/rest/blob/1b021a89a9c00b3d4cb5072b882963a430b340b1/src/lib/CDN.ts#L120-L129

`defaultAvatar`:
https://github.com/dirigeants/rest/blob/master/src/lib/CDN.ts#L43-L49

@vladfrangu

File: src/lib/caching/structures/User.ts [hunk 6]
```diff
@@ -138,6 +138,34 @@ export class User<T = Client> extends Structure<T> {
...
+	public avatarURL(options?: ImageURLOptions): string | null {
+		if (!this.avatar || !(this.client instanceof Client)) return null;
+		return this.client.api.cdn.userAvatar(this.id, this.avatar, options);
+	}
+
+	/**
+	 * Returns the users avatar url or the default discord avatar url if they don't have a avatar.
+	 * @param options The image size, format and other options.
+	 */
+	public displayAvatarURL(options?: ImageURLOptions): T extends Client ? string : null {
```
Comment (vladfrangu) [inline src/lib/caching/structures/User.ts:154]: I..really dislike the triple casting that is done in these files... Can't we do a generic, like `displayAvatarURL<K = T extends Client ? string : null>(blahBlah: optiowons): K` and cast as K?

File: src/lib/caching/structures/User.ts [hunk 7]
```diff
@@ -138,6 +138,34 @@ export class User<T = Client> extends Structure<T> {
...
+		if (!(this.client instanceof Client)) return null as T extends Client ? string : null;
+		return (this.avatar ?
+			this.avatarURL(options) :
+			this.client.api.cdn.defaultAvatar(Number(this.discriminator))) as T extends Client ? string : null;
+	}
+
+	/**
+	 * Returns the default discord avatar url for the user's discriminator.
+	 */
+	public defaultAvatarURL(): T extends Client ? string : null {
```
Comment (vladfrangu) [inline src/lib/caching/structures/User.ts:164]: Ditto

File: src/lib/caching/structures/User.ts [hunk 8]
```diff
@@ -138,6 +138,34 @@ export class User<T = Client> extends Structure<T> {
...
+
+	/**
+	 * Returns the users avatar url or the default discord avatar url if they don't have a avatar.
+	 * @param options The image size, format and other options.
+	 */
+	public displayAvatarURL(options?: ImageURLOptions): T extends Client ? string : null {
+		if (!(this.client instanceof Client)) return null as T extends Client ? string : null;
+		return (this.avatar ?
+			this.avatarURL(options) :
+			this.client.api.cdn.defaultAvatar(Number(this.discriminator))) as T extends Client ? string : null;
```
Comment (vladfrangu) [inline src/lib/caching/structures/User.ts:158]:
```suggestion
			this.defaultAvatarURL()) as T extends Client ? string : null;
```

File: src/lib/caching/structures/guilds/GuildMember.ts [hunk 9]
```diff
@@ -212,6 +212,22 @@ export class GuildMember extends Structure {
...
+	}
+
+	/**
+	 * Returns the users avatar url or the default discord avatar url if they don't have a avatar.
+	 * @param options The image size, format and other options.
+	 */
+	public displayAvatarURL(options?: ImageURLOptions): string {
+		return this.avatarURL(options) ?? this.client.api.cdn.defaultAvatar(Number(this.user?.discriminator));
+	}
+
```
Comment (vladfrangu) [inline src/lib/caching/structures/guilds/GuildMember.ts:230]: default here too thx

File: src/lib/caching/structures/guilds/GuildMember.ts [hunk 10]
```diff
@@ -212,6 +212,22 @@ export class GuildMember extends Structure {
 		return this;
 	}
 
+	/**
+	 * Returns the users avatar url.
+	 * @param options The image size, format and other options.
+	 */
+	public avatarURL(options?: ImageURLOptions): string | null {
+		return this.user?.avatar ? this.client.api.cdn.userAvatar(this.id, this.user.avatar, options) : null;
```
Comment (vladfrangu) [inline src/lib/caching/structures/guilds/GuildMember.ts:220]: Both this and display should return the user one, via `return this.user?.<methodName>(options) ?? null`

File: src/lib/caching/structures/guilds/GuildMember.ts [hunk 11]
```diff
@@ -212,6 +212,31 @@ export class GuildMember extends Structure {
 		return this;
 	}
 
+	/**
+	 * Returns the users avatar url.
+	 * @param options The image size, format and other options.
+	 */
+	public avatarURL(options?: ImageURLOptions): string | null {
```
Comment (bdistin) [inline src/lib/caching/structures/guilds/GuildMember.ts:219]: This is a useless alias, this should only exist on the user.

File: src/lib/caching/structures/guilds/GuildMember.ts [hunk 12]
```diff
@@ -212,6 +212,31 @@ export class GuildMember extends Structure {
...
+		return this.user?.avatarURL(options) ?? null;
+	}
+
+	/**
+	 * Returns the users avatar url or the default discord avatar url if they don't have a avatar.
+	 * @param options The image size, format and other options.
+	 */
+	public displayAvatarURL(options?: ImageURLOptions): string {
```
Comment (bdistin) [inline src/lib/caching/structures/guilds/GuildMember.ts:227]: ditto

File: src/lib/caching/structures/guilds/GuildMember.ts [hunk 13]
```diff
@@ -212,6 +212,31 @@ export class GuildMember extends Structure {
...
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		return this.user!.displayAvatarURL(options);
+	}
+
+	/**
+	 * Returns the default discord avatar url for the user's discriminator.
+	 */
+	public defaultAvatarURL(): string {
```
Comment (bdistin) [inline src/lib/caching/structures/guilds/GuildMember.ts:235]: ditto

File: src/lib/caching/structures/User.ts [hunk 14]
```diff
@@ -138,6 +138,34 @@ export class User<T = Client> extends Structure<T> {
...
+		if (!(this.client instanceof Client)) return null as unknown as K;
+		return (this.avatar ?
+			this.avatarURL(options) :
+			this.defaultAvatarURL()) as unknown as K;
+	}
+
+	/**
+	 * Returns the default discord avatar url for the user's discriminator.
+	 */
+	public defaultAvatarURL(): T extends Client ? string : null {
```
Comment (bdistin) [inline src/lib/caching/structures/User.ts:164]: If there isn't url options, this should be a getter.

**Pull request merged**